source/DistanceMap.cpp

```diff
@@ -260,25 +261,26 @@ void DistanceMap::Init(const Ship *ship)
 				}
 
 		// Bail out if the maximum number of systems is reached.
-		if(hyperspaceFuel && !Propagate(top, false))
+		if(hyperspaceFuel && !Propagate(top, false, ship))
 			break;
-		if(jumpFuel && !Propagate(top, true))
+		if(jumpFuel && !Propagate(top, true, ship))
 			break;
 	}
 }
 
 
 
 // Add the given links to the map. Return false if an end condition is hit.
-bool DistanceMap::Propagate(Edge edge, bool useJump)
+bool DistanceMap::Propagate(Edge edge, bool useJump, const Ship *ship)
 {
 	edge.fuel += (useJump ? jumpFuel : hyperspaceFuel);
 	for(const System *link : (useJump ? edge.next->JumpNeighbors(jumpRange) : edge.next->Links()))
 	{
 		// Find out whether we already have a better path to this system, and
 		// check whether this link can be traveled. If this route is being
 		// selected by the player, they are constrained to known routes.
-		if(HasBetter(*link, edge) || !CheckLink(*edge.next, *link, useJump))
+		if((ship && ship->IsRestrictedFrom (*link))
+				|| HasBetter(*link, edge) || !CheckLink(*edge.next, *link, useJump))
```

Comment:
This doesn't seem like the best place to put it; the proper place would IMO be CheckLink. And instead of passing the ship everywhere, add it as a member variable.

Reply:
~~everywhere as in one place? fine~~